File: android/app/src/main/java/com/exptrack/app/SMSBroadcastReceiver.java
package com.exptrack.app;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.provider.Telephony;
import android.telephony.SmsMessage;
import android.util.Log;
import android.widget.Toast;

public class SMSBroadcastReceiver extends BroadcastReceiver {
    private static final String TAG = "ExpTrack_SMS";

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.i(TAG, "onReceive called with action: " + (intent.getAction() != null ? intent.getAction() : "null"));
        
        try {
            if (Telephony.Sms.Intents.SMS_RECEIVED_ACTION.equals(intent.getAction())) {
                Log.d(TAG, "SMS_RECEIVED_ACTION matched");
                
                Bundle bundle = intent.getExtras();
                if (bundle != null) {
                    Object[] pdus = (Object[]) bundle.get("pdus");
                    String format = bundle.getString("format");
                    
                    if (pdus != null) {
                        Log.d(TAG, "Number of PDUs: " + pdus.length);
                        
                        for (Object pdu : pdus) {
                            SmsMessage message;
                            if (format != null && android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
                                message = SmsMessage.createFromPdu((byte[]) pdu, format);
                            } else {
                                message = SmsMessage.createFromPdu((byte[]) pdu);
                            }
                            
                            String messageBody = message.getMessageBody();
                            String sender = message.getOriginatingAddress();
                            
                            Log.i(TAG, "SMS Details:");
                            Log.i(TAG, "From: " + (sender != null ? sender : "unknown"));
                            Log.i(TAG, "Message: " + (messageBody != null ? messageBody : "empty"));
                            
                            // Show a toast for debugging
                            String toastMessage = "SMS From: " + sender + "\nMessage: " + messageBody;
                            Toast.makeText(context, toastMessage, Toast.LENGTH_LONG).show();
                            
                            // Check if it's a UPI message with more detailed logging
                            if (messageBody != null) {
                                boolean isUPI = messageBody.contains("UPI");
                                boolean isDebited = messageBody.contains("debited");
                                boolean isCredited = messageBody.contains("credited");
                                boolean isAccount = messageBody.contains("A/C");
                                
                                Log.d(TAG, "Message checks - UPI: " + isUPI + 
                                          ", Debited: " + isDebited + 
                                          ", Credited: " + isCredited + 
                                          ", A/C: " + isAccount);
                                
                                if (isUPI || isDebited || isCredited || isAccount) {
                                    Log.d(TAG, "UPI transaction SMS detected, sending to React Native");
                                    SMSListenerModule.emitMessageReceived(context, messageBody);
                                } else {
                                    Log.d(TAG, "Not a UPI message - ignoring");
                                }
                            } else {
                                Log.w(TAG, "Received null message body");
                            }
                        }
                    } else {
                        Log.w(TAG, "No PDUs found in intent");
                    }
                } else {
                    Log.w(TAG, "Null bundle received with SMS intent");
                }
            } else {
                Log.d(TAG, "Not an SMS_RECEIVED action");
            }
        } catch (Exception e) {
            Log.e(TAG, "Error processing SMS: " + e.getMessage());
            e.printStackTrace();
        }
    }
} 

File: android/app/src/main/java/com/exptrack/app/SMSBroadcastReceiver.kt
package com.exptrack.app

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.provider.Telephony
import android.util.Log
import android.content.pm.PackageManager
import com.facebook.react.ReactApplication
import com.facebook.react.bridge.ReactContext
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.facebook.react.bridge.Arguments

class SMSBroadcastReceiver : BroadcastReceiver() {
    private val TAG = "ExpTrack_SMS"

    override fun onReceive(context: Context?, intent: Intent?) {
        Log.i(TAG, "⭐ SMS RECEIVED - onReceive called ⭐")

        if (context == null || intent == null) {
            Log.e(TAG, "❌ Context or Intent is null")
            return
        }

        // Check if we have the necessary permissions
        if (context.checkSelfPermission(android.Manifest.permission.RECEIVE_SMS) != PackageManager.PERMISSION_GRANTED ||
            context.checkSelfPermission(android.Manifest.permission.READ_SMS) != PackageManager.PERMISSION_GRANTED) {
            Log.e(TAG, "❌ SMS permissions not granted")
            return
        }

        if (intent.action == Telephony.Sms.Intents.SMS_RECEIVED_ACTION) {
            Log.d(TAG, "📱 SMS_RECEIVED_ACTION matched")

            try {
                val messages = Telephony.Sms.Intents.getMessagesFromIntent(intent)
                Log.d(TAG, "📨 Processing ${messages.size} message parts")

                if (messages.isNotEmpty()) {
                    val smsMessage = messages[0]
                    val sender = smsMessage.originatingAddress ?: "Unknown"
                    val messageBody = smsMessage.messageBody ?: ""

                    Log.i(TAG, "📬 SMS Details Received:")
                    Log.i(TAG, "From: $sender")
                    Log.i(TAG, "Message: $messageBody")

                    // Check if it's a UPI transaction message
                    if (messageBody.contains("UPI", ignoreCase = true) || 
                        messageBody.contains("debited", ignoreCase = true) || 
                        messageBody.contains("credited", ignoreCase = true)) {
                        Log.d(TAG, "💰 UPI Transaction detected!")
                        
                        try {
                            // First try to emit to foreground app
                            emitToForeground(context, messageBody, sender)
                            
                            // Also start headless task for background processing
                            SMSHeadlessTask.startTask(context, messageBody, sender)
                            
                            Log.d(TAG, "✅ Message processing initiated")
                        } catch (e: Exception) {
                            Log.e(TAG, "❌ Error processing message: ${e.message}")
                            e.printStackTrace()
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ Error reading SMS: ${e.message}")
                e.printStackTrace()
            }
        }
    }

    private fun emitToForeground(context: Context, message: String, from: String) {
        try {
            val reactApp = context.applicationContext as ReactApplication
            val reactContext = reactApp.reactNativeHost.reactInstanceManager.currentReactContext

            if (reactContext != null && reactContext.hasActiveCatalystInstance()) {
                val params = Arguments.createMap().apply {
                    putString("message", message)
                    putString("from", from)
                    putLong("timestamp", System.currentTimeMillis())
                }

                reactContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit("onSMSReceived", params)

                Log.d(TAG, "✅ Message emitted to foreground app")
            } else {
                Log.d(TAG, "ℹ️ App not in foreground, using headless task only")
            }
        } catch (e: Exception) {
            Log.e(TAG, "❌ Error emitting to foreground: ${e.message}")
            e.printStackTrace()
        }
    }
} 